src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -634,5 +1079,86 @@ internal bool TryGetSpanOnHigherBuffer(SnapshotSpan snapshotSpan, ITextBuffer ta
             span = default;
             return false;
         }
+
+        private sealed class State
```

Comment:
being a class helped here.  thougths on making this immutable and just using replacement in the caller to represetnt states transitioning from one to another (a-la a state-machine), versus mutating the data in here?

Reply:
consider moving to another file .

Reply:
I'll keep it here for now to simplify the incremental review.

Reply:
It could work, but I have doubts that it would be significantly clearer. The parts of this type that make the states hard (e.g. reentrancy) are out of our control. Most of the meaningful changes will come from a replacement for this API.

Reply:
Filed #50767